build/ci.go

```diff
@@ -345,11 +349,88 @@ func downloadSpecTestFixtures(csdb *build.ChecksumDB, cachedir string) string {
 	return filepath.Join(cachedir, base)
 }
 
+// hashDir iterates all files under the top-level project directory
+// computing the hash of each file (excluding files within the tests
+// subrepo)
+func hashDir() (map[string]common.Hash, error) {
+	res := make(map[string]common.Hash)
+	err := filepath.WalkDir(".", func(path string, d os.DirEntry, err error) error {
+		if strings.HasPrefix(path, filepath.FromSlash("tests/testdata")) {
+			return nil
+		}
+		if d.IsDir() {
+			return nil
+		}
+		// open the file and hash it
+		f, err := os.OpenFile(path, os.O_RDONLY, 0666)
+		if err != nil {
+			return err
+		}
+		contents, err := io.ReadAll(f)
+		if err != nil {
+			return err
+		}
+		res[path] = common.Hash(sha256.Sum256(contents))
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+var cachedir = flag.String("cachedir", "./build/cache", "directory for caching golangci-lint/protoc binaries.")
+
+// doGenerate ensures that re-generating generated files does not cause
+// any mutations in the source file tree:  i.e. all generated files were
+// updated and committed.  Any stale generated files are updated.
+func doGenerate() {
+	tc := new(build.GoToolchain)
+
+	protocPath := downloadProtoc(*cachedir)
+
+	prevDir, err := os.Getwd()
+	if err != nil {
+		log.Fatalf("Getwd failed", "err", err)
+	}
+	if err := os.Chdir(os.TempDir()); err != nil {
```

Comment:
Just fyi, it's not necessary to do this, you can just set the directory the command runs in on the command itself. To do this, first create the command, the configure it, then run it with `build.MustRun`.

However, I also think what you're doing here generally isn't the right thing. We should download `protoc-gen-go` instead, in the same way we download `protoc` They have releases here: https://github.com/protocolbuffers/protobuf-go/releases/tag/v1.34.2

Reply:
> accounts/usbwallet/trezor/trezor.go:45: running "protoc": exec: "protoc": cannot run executable found relative to current directory

I think we will have to place `protoc-gen-go` outside `build` and under a temp directory (randomly-generated dir name under top-level temp directory).

Reply:
The error probably just means you need to put an absolute directory in into PATH?

Reply:
wait this refers to `protoc` being on a relative path.